### PR TITLE
quartus-prime-lite: 20.1.1.720 -> 22.1std.2.922

### PR DESCRIPTION
**Pull request by haskelious**
## Description of changes

Update Quartus Prime Lite to 22.1.2.
Homepage of this version is https://www.intel.com/content/www/us/en/software-kit/785085/intel-quartus-prime-lite-edition-design-software-version-22-1-2-for-linux.html

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

PRs related to this one (possibly in conflict):
- https://github.com/NixOS/nixpkgs/pull/204401

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/editors/quartus-prime/default.nix [hunk 1]
```diff
@@ -98,10 +103,10 @@ in buildFHSEnv rec {
     # https://community.intel.com/t5/Intel-FPGA-Software-Installation/Running-Quartus-Prime-Standard-on-WSL-crashes-in-libudev-so/m-p/1189032
     #
     # But, as can be seen in the above resource, LD_PRELOADing libudev breaks
-    # compiling encrypted device libraries in ModelSim (with error
+    # compiling encrypted device libraries in Questa (with error
     # `(vlog-2163) Macro `<protected> is undefined.`), so only use LD_PRELOAD
     # for non-ModelSim wrappers.
```
Comment (bjornfor) [inline pkgs/applications/editors/quartus-prime/default.nix:108]: non-ModelSim -> non-Questa

**Review by bjornfor — COMMENTED** [review]
Tested:

* `quartus` builds and runs (GUI opens).
* `nix-build -A quartus-prime-lite.tests` pass (so `questa` should work).

NOTE: `vsim` now requires a license to run, but apparently one can run `vcom` and `vlog` without a license.

**Comment by bjornfor** [comment]
This reminded me that we're lacking at test for quartus. So I made https://github.com/NixOS/nixpkgs/pull/277774.

File: pkgs/applications/editors/quartus-prime/default.nix (hunk 1 above)
Comment (bjornfor) [reply pkgs/applications/editors/quartus-prime/default.nix:108]: If you fix that s/ModelSim/Questa above I'll approve.

Comment (haskelious) [reply pkgs/applications/editors/quartus-prime/default.nix]: @bjornfor I pushed the requested commit

**Comment by haskelious** [comment]
Indeed, questa FPGA starter edition does require a license, free to get from https://licensing.intel.com/psg/s/
I tested questa (`vsim`) with a license and it runs successfully.

**Review by bjornfor — APPROVED** [review]

**Comment by haskelious** [comment]
It looks like the Help links do not manage to open the local web browser due to library errors. This is caused by the value of LD_LIBRARY_PATH in the environment used when quartus opens the browser. I do not know how to fix this in the package.
As a workaround, I used a custom script that contains
```
env LD_LIBRARY_PATH= xdg-open $*
```
and set the path to this script in Tools >> Options >> General >> Internet Connectivity >> Use custom web browser

**Comment by bjornfor** [comment]
I think it is quartus itself (not nixpkgs wrappers) that set LD_LIBRARY_PATH, so it might require patching to fix.

Can xdg.portal.xdgOpenUsePortal be of help here (on NixOS at least)?

**Comment by Herschenglime** [comment]
Hi! Do you also plan on packaging quartus 23.1? I got it to build with an overlay on top of the old 20.1 version in the stable repo with some rehashing, and changes to the url paths as seen here:

```nix
# This file defines overlays
{ inputs, ... }: {
  # This one contains whatever you want to overlay
  # You can change versions, add patches, set compilation flags, anything really.
  # https://nixos.wiki/wiki/Overlays
  modifications = final: prev: {
    # use prev.<something> to access atributes of original package
    quartus-prime-lite = let
      supportedDevices = [
        "Arria II"
        "Cyclone V"
        "Cyclone IV"
        "Cyclone 10 LP"
        "MAX II/V"
        "MAX 10 FPGA"
      ];
      deviceIds = {
        "Arria II" = "arria_lite";
        "Cyclone V" = "cyclonev";
        "Cyclone IV" = "cyclone";
        "Cyclone 10 LP" = "cyclone10lp";
        "MAX II/V" = "max";
        "MAX 10 FPGA" = "max10";
      };

      supportedDeviceIds = assert prev.lib.assertMsg
        (prev.lib.all (name: prev.lib.hasAttr name deviceIds) supportedDevices)
        "Supported devices are: ${
          prev.lib.concatStringsSep ", " (prev.lib.attrNames deviceIds)
        }";
        prev.lib.listToAttrs (map (name: {
          inherit name;
          value = deviceIds.${name};
        }) supportedDevices);

      componentHashes = {
        "arria_lite" = "sha256-FH/UrPnbGCy94GfTR+aw8Lq3X66vw8FWrRutxWyt6Tk=";
        "cyclone" = "sha256-1z84CmdXsa0Mfd8Lip4yg38Gz4XR0P0CA5QiriQMNtY=";
        "cyclone10lp" = "sha256-sueWxLzvcbvEyw5cDpcJZA+vZoW70zhPNBA8CUOoUs4=";
        "cyclonev" = "sha256-77zQXr5SFJH6ZyCmOFeY5aAAoY6NQfg1HRw7DzESaYM=";
        "max" = "sha256-az1I+U2gPp3urYHREvxmFel5NG4Qah1zjF9aPP7/WhM=";
        "max10" = "sha256-4iRwfg8e1X7ZOcJMTfpOYavYmQ90LgqLu6S3fw7BFDc=";
      };

      version = "23.1.0.991";
      versionWithStd =
        "23.1std.0.991"; # there might be a cleaner way to do this but I was in a hurry

      # new download link: https://downloads.intel.com/akdlm/software/acdsinst/23.1std/991/ib_installers/cyclone-23.1std.0.991.qdz
      download = { name, sha256 }:
        prev.fetchurl {
          inherit name sha256;
          # e.g. "20.1.1.720" -> "20.1std.1/720"

          # modified
          url = "https://downloads.intel.com/akdlm/software/acdsinst/${
              prev.lib.versions.majorMinor version
            }std/${
              prev.lib.elemAt (prev.lib.splitVersion version) 3
            }/ib_installers/${name}";
        };
    in prev.quartus-prime-lite.overrideAttrs (oldAttrs: rec {
      src = map download ([
        {
          name = "QuartusLiteSetup-${versionWithStd}-linux.run";
          sha256 = "sha256-EKXf6WQztmVi/641LxvSdKXyTdL+363bPIC8bMpq5NU=";
        }
        {
          name = "QuestaSetup-${versionWithStd}-linux.run";
          sha256 = "sha256-s0cBn3Ucq2/5NOEhmP4pg0oUHI5kzeGijMRtMuH1NDk=";
        }
      ] ++ (map (id: {
        name = "${id}-${versionWithStd}.qdz"; # modified
        sha256 = prev.lib.getAttr id componentHashes;
      }) (prev.lib.attrValues supportedDeviceIds)));
    }); # last line of quartus
  };
}
```

However, for some reason, the updated package wouldn't actually kick in on my system.

I'm not sure if the changes here would work with this particular commit, but hopefully it at least saves the work of getting the new hashes.

**Comment by haskelious** [comment]
I built also a 23.1 version locally. I can try to submit a PR for that also.

**Comment by haskelious** [comment]
@Herschenglime see https://github.com/NixOS/nixpkgs/pull/280121 if you'd like to test

**Pull request merged**